README.md
# Franklin Domainkey Provider

> Create domainkeys for your domain by completing a DNS or HTTP challenge

## Status
[![codecov](https://img.shields.io/codecov/c/github/adobe/franklin-domainkey-provider.svg)](https://codecov.io/gh/adobe/franklin-domainkey-provider)
[![CircleCI](https://img.shields.io/circleci/project/github/adobe/franklin-domainkey-provider.svg)](https://circleci.com/gh/adobe/franklin-domainkey-provider)
[![GitHub license](https://img.shields.io/github/license/adobe/franklin-domainkey-provider.svg)](https://github.com/adobe/franklin-domainkey-provider/blob/main/LICENSE.txt)
[![GitHub issues](https://img.shields.io/github/issues/adobe/franklin-domainkey-provider.svg)](https://github.com/adobe/franklin-domainkey-provider/issues)
[![LGTM Code Quality Grade: JavaScript](https://img.shields.io/lgtm/grade/javascript/g/adobe/franklin-domainkey-provider.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/adobe/franklin-domainkey-provider)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)

## Usage

Start by calling the service

```bash
$ curl https://eynvwoxb7l.execute-api.us-east-1.amazonaws.com/helix-services/domainkey-provider/v1/
```

It will tell you that it needs a `domain` parameter, so we try again

```bash
$ curl -X POST -d domain=example.com https://eynvwoxb7l.execute-api.us-east-1.amazonaws.com/helix-services/domainkey-provider/v1/
```

This will return instruction on setting completing the callenge. The response
contains a domain key that will be a UUID like `f4a5cb7f-adac-450c-919f-a12b13cec116`
as well as a challenge that is a hash of the domain key and your domain like
`4a159285c173d7ac98a3e20c746b46d191ea14dd53214b42a2f6ed36f7d2aeb7`

There are now two ways to complete the challenge, either by setting a DNS record, or by providing an HTTP response.

### DNS Challenge

Create a TXT record for `_rum_-challenge.example.com` with the value of the challenge. 
You can verify that the record has been set using `dig`

```bash
$ dig TXT _rum-challenge.example.com
```

Once the record is set, you can call the service again to verify that the challenge
has been completed and start issuing domain keys.

```bash
curl -X POST -d "domain=example.com&domainkey=f4a5cb7f-adac-450c-919f-a12b13cec116" https://eynvwoxb7l.execute-api.us-east-1.amazonaws.com/helix-services/domainkey-provider/v1/
```

If the domain key has been verified and activated, you will see a response status of
201. If the domain key has not been verified, you will see a response status of 403.

### HTTP Challenge

If you are using `*.hlx.live` as your origin, your CDN is already set up to complete the challenge.

If not, ensure that at `https://example.com/_rum-challenge` a response with the status code `204` and following response headers is created
- `x-rum-challenge`: the challenge value or a list of challenge values separated by a space

You can verify that the challenge has been set using `curl`

```bash
$ curl -s -I https://example.com/_rum-challenge | grep x-rum-challenge
x-rum-challenge: 4a159285c173d7ac98a3e20c746b46d191ea14dd53214b42a2f6ed36f7d2aeb7
```

Once the challenge has been set, you can call the service again to verify that the challenge has been completed and start issuing domain keys.

```bash
$ curl -X POST -d "domain=example.com&domainkey=f4a5cb7f-adac-450c-919f-a12b13cec116" https://eynvwoxb7l.execute-api.us-east-1.amazonaws.com/helix-services/domainkey-provider/v1/
```

## Development

### Deploying Franklin Domainkey Provider

All commits to main that pass the testing will be deployed automatically. All commits to branches that will pass the testing will get commited as `/helix-services/service@ci<num>` and tagged with the CI build number.


CHANGELOG.md
# [1.1.0](https://github.com/adobe/franklin-domainkey-provider/compare/v1.0.4...v1.1.0) (2023-11-02)


### Bug Fixes

* **http:** handle http timeouts properly ([0366aba](https://github.com/adobe/franklin-domainkey-provider/commit/0366abae482d6676ec483c3b6943e521382fd779))
* **http:** use correct fetch api ([d68e368](https://github.com/adobe/franklin-domainkey-provider/commit/d68e36830846bce4eea540aeefe976ecefb145a4))


### Features

* **http:** add instructions for HTTP challenge in initial response ([9fe17eb](https://github.com/adobe/franklin-domainkey-provider/commit/9fe17ebb8a57cb54c9fc1f4ec1b465080208f324))
* **index:** implement http challenge ([85a9d41](https://github.com/adobe/franklin-domainkey-provider/commit/85a9d41aa1465edc01ac32c59f2bb9548be38a53))
* **index:** support both HTTP and DNS validation ([5b25c4e](https://github.com/adobe/franklin-domainkey-provider/commit/5b25c4e85597855704ebb893a69b1f3d67eeb572))

## [1.0.4](https://github.com/adobe/franklin-domainkey-provider/compare/v1.0.3...v1.0.4) (2023-10-30)


### Bug Fixes

* **deps:** update dependency @adobe/fetch to v4.1.1 ([#34](https://github.com/adobe/franklin-domainkey-provider/issues/34)) ([6c6e046](https://github.com/adobe/franklin-domainkey-provider/commit/6c6e046671368fa0c16f2081bc2985799a0f3dc6))

## [1.0.3](https://github.com/adobe/franklin-domainkey-provider/compare/v1.0.2...v1.0.3) (2023-10-24)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal-logger to v3.0.12 ([#32](https://github.com/adobe/franklin-domainkey-provider/issues/32)) ([329fc3e](https://github.com/adobe/franklin-domainkey-provider/commit/329fc3e08c5357177e10e26158a83ab4f0ace67c))

## [1.0.2](https://github.com/adobe/franklin-domainkey-provider/compare/v1.0.1...v1.0.2) (2023-10-07)


### Bug Fixes

* **deps:** update dependency @adobe/fetch to v4.1.0 ([#14](https://github.com/adobe/franklin-domainkey-provider/issues/14)) ([3443709](https://github.com/adobe/franklin-domainkey-provider/commit/344370992ddab24af15f27d030a181635a9a1e7d))

## [1.0.1](https://github.com/adobe/franklin-domainkey-provider/compare/v1.0.0...v1.0.1) (2023-07-28)


### Bug Fixes

* **index:** always read the response ([27b3617](https://github.com/adobe/franklin-domainkey-provider/commit/27b3617a1003371f3e0e299e2350275644a89ad0))
* **index:** better error handling ([bf7d2f6](https://github.com/adobe/franklin-domainkey-provider/commit/bf7d2f6153a038c4965a958f16309ccd716dcf8a))
* key is results, not result ([a9ff8fa](https://github.com/adobe/franklin-domainkey-provider/commit/a9ff8fa6201991e0f8088d47611e3b30a718d96f))

# 1.0.0 (2023-07-28)


### Features

* initial implementation ([dd20a05](https://github.com/adobe/franklin-domainkey-provider/commit/dd20a05fccaee85437822abc00d315d06063f09e))


test/index.test.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */
import assert from 'assert';
import { Request } from '@adobe/fetch';
import { main } from '../src/index.js';

describe('Index Tests', () => {
  it('index rejects to run if no domain key is available', async () => {
    const result = await main(new Request('https://localhost/'), {
      env: {},
    });
    assert.equal(result.status, 500);
    assert.strictEqual(await result.text(), 'No HELIX_RUN_QUERY_DOMAIN_KEY set. This is a configuration error');
  });

  it('index requires a domain', async () => {
    const result = await main(new Request('https://localhost/'), {
      env: {
        HELIX_RUN_QUERY_DOMAIN_KEY: 'foo',
      },
    });
    assert.equal(result.status, 400);
    assert.strictEqual(await result.text(), 'No domain specified');
  });

  it('index generates a domain key if not specified', async () => {
    const result = await main(new Request('https://localhost/', {
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      method: 'POST',
      body: 'domain=example.com',
    }), {
      env: {
        HELIX_RUN_QUERY_DOMAIN_KEY: 'foo',
      },
    });
    assert.equal(result.status, 404);
    const newkey = result.headers.get('x-domainkey');
    const txt = await result.text();
    assert(txt.indexOf(newkey) >= 0);
  });

  it('index returns 404 if text record is not set', async () => {
    const result = await main(new Request('https://localhost/', {
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      method: 'POST',
      body: 'domain=example.com&domainkey=foo',
    }), {
      env: {
        HELIX_RUN_QUERY_DOMAIN_KEY: 'foo',
      },
      logger: console,
    });
    assert.equal(result.status, 404);
  });

  it('index returns 403 if text record is set but wrong', async () => {
    const result = await main(new Request('https://localhost/', {
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      method: 'POST',
      body: 'domain=johansminecraft.club&domainkey=bar',
    }), {
      env: {
        HELIX_RUN_QUERY_DOMAIN_KEY: 'foo',
      },
      logger: console,
    });
    assert.equal(result.status, 403);
  });

  it('index returns 503 if rotating the domainkey failed for backend reasons', async () => {
    const result = await main(new Request('https://localhost/', {
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      method: 'POST',
      body: 'domain=johansminecraft.club&domainkey=foo',
    }), {
      env: {
        HELIX_RUN_QUERY_DOMAIN_KEY: 'baz', // this key is wrong on purpose, so that the update won't go through
      },
      logger: console,
    });
    assert.equal(result.status, 503);
  }).timeout(50000);

  it('index returns 404 if no rum challenge header is returned', async () => {
    const result = await main(new Request('https://localhost/', {
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      method: 'POST',
      body: 'domain=eomfpog1mphtbgh.m.pipedream.net&domainkey=foo',
    }), {
      env: {
        HELIX_RUN_QUERY_DOMAIN_KEY: 'baz',
      },
      logger: console,
    });
    assert.equal(result.status, 404);
  });

  it('index returns 201 if correct rum challenge header is returned', async () => {
    const result = await main(new Request('https://localhost/', {
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      method: 'POST',
      body: 'domain=minivelos.bike&domainkey=foo',
    }), {
      env: {
        HELIX_RUN_QUERY_DOMAIN_KEY: 'baz',
      },
      logger: console,
    });
    assert.equal(result.status, 503); // update failed intentionally
  }).timeout(50000);

  it('index returns 403 if wrong rum challenge header is returned', async () => {
    const result = await main(new Request('https://localhost/', {
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      method: 'POST',
      body: 'domain=forward.minivelos.workers.dev&domainkey=foo',
    }), {
      env: {
        HELIX_RUN_QUERY_DOMAIN_KEY: 'baz',
      },
      logger: console,
    });
    assert.equal(result.status, 403);
  });

  it('index returns 201 if rotating the domainkey worked as expected', async () => {
    const result = await main(new Request('https://localhost/', {
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      method: 'POST',
      body: 'domain=johansminecraft.club&domainkey=foo',
    }), {
      env: {
        HELIX_RUN_QUERY_DOMAIN_KEY: process.env.TEST_DOMAINKEY, // this one is real,
        // but only for one domain
      },
      logger: console,
    });
    assert.equal(result.status, 201, await result.text());
  }).timeout(50000);
});


test/post-deploy.test.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */
import assert from 'assert';
import { h1NoCache } from '@adobe/fetch';
import { createTargets } from './post-deploy-utils.js';

createTargets().forEach((target) => {
  describe(`Post-Deploy Tests (${target.title()})`, () => {
    const fetchContext = h1NoCache();
    const { fetch } = fetchContext;

    afterEach(() => {
      fetchContext.reset();
    });

    it(`returns the status of the function at ${target.url('/_status_check/healthcheck.json')}`, async () => {
      const url = target.url('/_status_check/healthcheck.json');
      const res = await fetch(url);
      assert.strictEqual(res.status, 200);
      const json = await res.json();
      delete json.process;
      delete json.response_time;
      // status returns 0.0.0+ci123 for ci versions
      const version = target.version.startsWith('ci')
        ? `0.0.0+${target.version}`
        : target.version;
      assert.deepStrictEqual(json, {
        status: 'OK',
        version,
      });
    }).timeout(50000);

    it(`invokes the function at ${target.url('/')}`, async () => {
      const res = await fetch(target.url('/'));
      assert.strictEqual(res.status, 400);
    }).timeout(50000);
  });
});


src/dev.js
import dotenv from 'dotenv';
import path from 'path';
dotenv.config({ path: path.resolve(process.cwd(), `.env.default`) });

import http from 'http';
import { run } from './run.js';
import { parseQuery } from './utils/query.js';

const DEV_PORT = process.env.DEV_PORT || 8080;
const HELIX_RUN_QUERY_DOMAIN_KEY = process.env.HELIX_RUN_QUERY_DOMAIN_KEY || '';

http.createServer(async function (req, res) {
    var query = '';
    req.on('readable', function() {
        const readPayload = req.read();
        if (readPayload) {
            query += readPayload;
        }
    });
    req.on('end', async () => {
        const context = {
            env: {
                HELIX_RUN_QUERY_DOMAIN_KEY,
            },
            data: JSON.stringify(parseQuery(query))
        };

        const request = {
            url: req.headers.host + req.url,
            method: req.method,
            headers: req.headers,
            query
        };

        const response = await run(request, context);
        const resolved = await response?.text();
    
        res.writeHead(response['status'], response['headers']);
        res.end(JSON.stringify(resolved));
    });
  }).listen(DEV_PORT, 'localhost', () => {
    console.log(`Server is running on http://localhost:${DEV_PORT}`);
});



src/run.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
import wrap from '@adobe/helix-shared-wrap';
import { logger } from '@adobe/helix-universal-logger';
import { helixStatus } from '@adobe/helix-status';
import { Response, fetch } from '@adobe/fetch';
import bodyData from '@adobe/helix-shared-body-data';
import { randomUUID, createHash } from 'crypto';
import { resolveTxt } from 'dns';
import { promisify } from 'util';

const resolveTxtAsync = promisify(resolveTxt);

function hashMe(domain, domainkey) {
  const hash = createHash('sha256');
  hash.update(domain);
  hash.update(domainkey);
  return hash.digest('hex');
}

/**
 * This is the main function
 * @param {Request} request the request object (see fetch api)
 * @param {UniversalContext} context the context of the universal serverless function
 * @returns {Response} a response
 */
export async function run(request, context) {
  const { HELIX_RUN_QUERY_DOMAIN_KEY } = context.env;
  const { data } = context;
  const { domain, domainkey } = data;
  if (!HELIX_RUN_QUERY_DOMAIN_KEY) {
    return new Response('No HELIX_RUN_QUERY_DOMAIN_KEY set. This is a configuration error', {
      status: 500,
    });
  }
  if (!domain) {
    return new Response('No domain specified', {
      status: 400,
    });
  }
  if (!domainkey) {
    // the domain key is not set, so generate a new key (a UUID) for the domain and
    // provide instructions on how to set up DNS validation
    const newkey = randomUUID();
    const hash = hashMe(domain, newkey);
    const currentURL = new URL(request.url);
    currentURL.search = '';

    const instructions = `Please add a TXT record for _rum-challenge.${domain} with the value ${hash}. Once
the record is added, you can verify that it is set up correctly by making a POST request to this URL including
the domain and domainkey parameters. For example:

curl -X POST -d "domain=${domain}&domainkey=${newkey}" ${currentURL}
`;
    return new Response(instructions, {
      status: 404,
      headers: {
        'Content-Type': 'text/plain',
        'x-error': 'domainkey not set',
        'x-domainkey': newkey,
      },
    });
  }
  // the domain key is set, so verify that the TXT record is set up correctly
  const hash = hashMe(domain, domainkey);
  const txt = `_rum-challenge.${domain}`;
  let txtrecords = [];
  try {
    txtrecords = await resolveTxtAsync(txt);
  } catch (e) {
    //context.logger.error(`Error while resolving TXT record for ${txt}: ${e.message}`);
  }
  if (txtrecords.length === 0) {
    return new Response(`No TXT record found for ${txt}`, {
      status: 404,
      headers: {
        'x-error': 'TXT record not found',
      },
    });
  }
  if (txtrecords[0].indexOf(hash) === -1) {
    return new Response(`TXT record for ${txt} does not contain ${hash}`, {
      status: 403,
      headers: {
        'x-error': 'TXT record does not match',
      },
    });
  }
  // create new domain key by making API request
  const endpoint = new URL('https://helix-pages.anywhere.run/helix-services/run-query@v3/rotate-domainkeys');
  const body = {
    url: domain,
    newkey: domainkey,
    domainkey: HELIX_RUN_QUERY_DOMAIN_KEY,
    readonly: false,
  };
  const res = await fetch(endpoint, {
    method: 'POST',
    body: JSON.stringify(body),
    headers: {
      'Content-Type': 'application/json',
    },
  });
  const json = await res.json();
  if (!res.ok || json.results.data[0].key !== domainkey) {
    return new Response(`Error while rotating domain keys: ${res.statusText}`, {
      status: 503,
    });
  }
  const confirmedkey = json.results.data[0].key;
  return new Response(`TXT record for ${txt} contains ${hash}, you can now use the domainkey ${confirmedkey}`, {
    status: 201,
  });
}

src/index.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
import wrap from '@adobe/helix-shared-wrap';
import { logger } from '@adobe/helix-universal-logger';
import { helixStatus } from '@adobe/helix-status';
import {
  Response,
  h1NoCache, AbortController, AbortError,
} from '@adobe/fetch';
import bodyData from '@adobe/helix-shared-body-data';
import { randomUUID, createHash } from 'crypto';
import { resolveTxt } from 'dns';
import { promisify } from 'util';

const { fetch } = h1NoCache({
  userAgent: 'franklin-domainkey-provider/1.0',
});

const resolveTxtAsync = promisify(resolveTxt);

function hashMe(domain, domainkey) {
  const hash = createHash('sha256');
  hash.update(domain);
  hash.update(domainkey);
  return hash.digest('hex');
}

async function validateDNS(domain, context, hash, confirmedkey) {
  const txt = `_rum-challenge.${domain}`;
  let txtrecords = [];
  try {
    txtrecords = await resolveTxtAsync(txt);
  } catch (e) {
    context.logger.error(`Error while resolving TXT record for ${txt}: ${e.message}`);
  }
  if (txtrecords.length === 0) {
    return new Response(`No TXT record found for ${txt}`, {
      status: 404,
      headers: {
        'x-error': 'TXT record not found',
      },
    });
  }
  if (txtrecords[0].indexOf(hash) === -1) {
    return new Response(`TXT record for ${txt} does not contain ${hash}`, {
      status: 403,
      headers: {
        'x-error': 'TXT record does not match',
      },
    });
  }
  return new Response(`TXT record for ${txt} contains ${hash}, you can now use the domainkey ${confirmedkey}`, {
    status: 201,
  });
}
/**
 * HTTP challenge validation makes a request to https://${domain}/_rum-challenge and expects
 * a 204 response with an x-rum-challenge header containing the hash.
 * @param {string} domain
 * @param {Context} context
 * @param {string} hash
 * @param {string} confirmedkey
 */
async function validateHTTP(domain, _context, hash, confirmedkey) {
  _context.logger.log('validating HTTP challenge', domain);
  const controller = new AbortController();
  const timerId = setTimeout(() => controller.abort(), 1000);
  const { signal } = controller;
  try {
    const res = await fetch(`https://${domain}/_rum-challenge`, {
      method: 'OPTIONS',
      signal,
    });
    // always read the body, otherwise the connection is not closed
    await res.text();
    if (res.status !== 204) {
      return new Response(`Error while validating HTTP challenge: ${res.statusText} is not a valid 204 status`, {
        status: 503,
      });
    }
    const challenge = res.headers.get('x-rum-challenge');
    if (!challenge) {
      return new Response('Error while validating HTTP challenge: no x-rum-challenge header set', {
        status: 404,
      });
    }
    const challenges = challenge.split(' ');
    if (challenges.find((c) => c === hash)) {
      return new Response(`HTTP challenge for https://${domain}/_rum-challenge contains ${hash}, you can now use the domainkey ${confirmedkey}`, {
        status: 201,
      });
    }
    return new Response(`HTTP challenge for https://${domain}/_rum-challenge does not contain ${hash}`, {
      status: 403,
      headers: {
        'x-error': 'HTTP challenge does not match',
      },
    });
  } catch (e) {
    if (e instanceof AbortError) {
      return new Response(`Timeout while validating HTTP challenge: ${e.message}`, {
        status: 504,
      });
    }
    /* c8 ignore start */
    return new Response(`Error while validating HTTP challenge: ${e.message}`, {
      status: 503,
    });
  } finally {
    clearTimeout(timerId);
  }
  /* c8 ignore stop */
}

/**
 * This is the main function
 * @param {Request} request the request object (see fetch api)
 * @param {UniversalContext} context the context of the universal serverless function
 * @returns {Response} a response
 */
async function run(request, context) {
  const { HELIX_RUN_QUERY_DOMAIN_KEY } = context.env;
  const { data } = context;
  const { domain, domainkey } = data;
  if (!HELIX_RUN_QUERY_DOMAIN_KEY) {
    return new Response('No HELIX_RUN_QUERY_DOMAIN_KEY set. This is a configuration error', {
      status: 500,
    });
  }
  if (!domain) {
    return new Response('No domain specified', {
      status: 400,
    });
  }
  if (!domainkey) {
    // the domain key is not set, so generate a new key (a UUID) for the domain and
    // provide instructions on how to set up DNS validation
    const newkey = randomUUID();
    const hash = hashMe(domain, newkey);
    const currentURL = new URL(request.url);
    currentURL.search = '';

    const instructions = `Please add a TXT record for _rum-challenge.${domain} with the value ${hash}. Once
the record is added, you can verify that it is set up correctly by making a POST request to this URL including
the domain and domainkey parameters. For example:

curl -X POST -d "domain=${domain}&domainkey=${newkey}" ${currentURL}

Alternatively, use the HTTP challenge and provide a resource at https://${domain}/_rum-challenge that responds to
an OPTIONS request with a 204 status code and following headers:
- x-rum-challenge: ${hash}

If you are using *.hlx.live as your CDN origin, these headers will be added automatically.
`;
    return new Response(instructions, {
      status: 404,
      headers: {
        'Content-Type': 'text/plain',
        'x-error': 'domainkey not set',
        'x-domainkey': newkey,
      },
    });
  }
  // the domain key is set, so verify that the TXT record is set up correctly
  const hash = hashMe(domain, domainkey);
  const [response] = (await Promise.all([
    validateDNS(domain, context, hash, domainkey),
    validateHTTP(domain, context, hash, domainkey),
  ]))
    // sort responses by status code, so that the first one is the best one
    .sort((a, b) => a.status - b.status);
  if (response.status !== 201) {
    // if the validation failed, return the response
    return response;
  }

  // create new domain key by making API request
  const endpoint = new URL('https://helix-pages.anywhere.run/helix-services/run-query@v3/rotate-domainkeys');
  const body = {
    url: domain,
    newkey: domainkey,
    domainkey: HELIX_RUN_QUERY_DOMAIN_KEY,
    readonly: false,
  };
  const res = await fetch(endpoint, {
    method: 'POST',
    body: JSON.stringify(body),
    headers: {
      'Content-Type': 'application/json',
    },
  });
  const json = await res.json();
  if (!res.ok || json.results.data[0].key !== domainkey) {
    return new Response(`Error while rotating domain keys: ${res.statusText}`, {
      status: 503,
    });
  }
  return response;
}

export const main = wrap(run)
  .with(helixStatus)
  .with(logger.trace)
  .with(logger)
  .with(bodyData);
